services/rfq/api/rest/server.go

```diff
@@ -360,13 +360,22 @@ func (r *QuoterAPIServer) checkRoleParallel(c *gin.Context, destChainID uint32)
 	var v1Addr, v2Addr common.Address
 	var v1Err, v2Err error
 
+	quoterRole := crypto.Keccak256Hash([]byte("QUOTER_ROLE"))
+	relayerRole := crypto.Keccak256Hash([]byte("RELAYER_ROLE"))
 	g.Go(func() error {
-		v1Addr, v1Err = r.checkRole(c, destChainID, true)
+		v1Addr, v1Err = r.checkRole(c, destChainID, true, quoterRole)
+		return v1Err
+	})
+	g.Go(func() error {
+		v1Addr, v1Err = r.checkRole(c, destChainID, true, relayerRole)
 		return v1Err
 	})
```

Comment:
The cache is contract-specific, but not binded to a role. So what likely to happen here is that we check for two different roles in the same contract, one check will result in the actual RPC call, while the second one will just reuse the cached value. Since we don't really need to check multiple roles in the same contract, I suggest to leave the role cache as is. Instead, let's only check for `RELAYER_ROLE` in V1 and for `QUOTER_ROLE` in V2. This also helps with the future cleanup once the V1 is fully deprecated.

Reply:
This could also lead to race conditions between two go routines